paddle/fluid/operators/lstsq_op.cu

```diff
@@ -0,0 +1,194 @@
+
+namespace paddle {
+namespace operators {
+
+using paddle::framework::Tensor;
+
+template <typename DeviceContext, typename T>
+class LstsqCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor& x = *context.Input<Tensor>("X");
+    const Tensor& y = *context.Input<Tensor>("Y");
+    auto* solution = context.Output<Tensor>("Solution");
+
+    auto dito =
+        math::DeviceIndependenceTensorOperations<platform::CUDADeviceContext,
+                                                 T>(context);
+    auto& dev_ctx =
+        context.template device_context<platform::CUDADeviceContext>();
+
+    auto x_dims = x.dims();
+    auto y_dims = y.dims();
+    int dim_size = x_dims.size();
+    int m = x_dims[dim_size - 2];
+    int n = x_dims[dim_size - 1];
+    int m_y = y_dims[dim_size - 2];
+    int n_y = y_dims[dim_size - 1];
+    int min_mn = std::min(m, n);
+    int max_mn = std::max(m, n);
+    int k = min_mn;
+
+    int x_stride = MatrixStride(x);
+    int y_stride = MatrixStride(y);
+    int tau_stride = min_mn;
+    int batch_count = BatchCount(x);
+
+    Tensor new_x, new_y;
+    new_x.mutable_data<T>(context.GetPlace(),
+                          size_t(batch_count * m * n * sizeof(T)));
+    new_y.mutable_data<T>(context.GetPlace(),
+                          size_t(batch_count * m_y * n_y * sizeof(T)));
+    framework::TensorCopy(x, context.GetPlace(), &new_x);
+    framework::TensorCopy(y, context.GetPlace(), &new_y);
+
+    // Prepare tau
+    auto tau_dims_vec = framework::vectorize<int>(x_dims);
+    tau_dims_vec.pop_back();
+    tau_dims_vec[tau_dims_vec.size() - 1] = min_mn;
+    Tensor tau = dito.Fill(tau_dims_vec, 0);
+    auto tau_data = tau.mutable_data<T>(context.GetPlace());
+
+    auto tmp_x = dito.Transpose(new_x);
+    auto tmp_y = dito.Transpose(new_y);
+    framework::TensorCopy(tmp_x, new_x.place(), &new_x);
+    framework::TensorCopy(tmp_y, new_y.place(), &new_y);
+
+    auto x_data = new_x.mutable_data<T>(context.GetPlace());
+    auto y_data = new_y.mutable_data<T>(context.GetPlace());
+
+    // step 1, compute QR factorization using geqrf
+    BatchedGeqrf<DeviceContext, T>(dev_ctx, batch_count, m, n, x_data, m,
+                                   tau_data, x_stride, tau_stride);
+
+    // Step 2, B <- Q^H B
+    BatchedOrmqr<DeviceContext, T>(dev_ctx, true, true, batch_count, m, n, k,
+                                   x_data, x_stride, tau_data, tau_stride,
+                                   y_data, y_stride);
+
+    auto trans_r = dito.Transpose(new_x);
+    auto trans_b = dito.Transpose(new_y);
+    auto slice_r = dito.Slice(trans_r, {-2}, {0}, {min_mn});
+    auto slice_b = dito.Slice(trans_b, {-2}, {0}, {min_mn});
+    auto tmp_r = dito.TrilTriu(slice_r, 0, false);
+    framework::TensorCopy(tmp_r, new_x.place(), &new_x);
+    framework::TensorCopy(slice_b, new_y.place(), &new_y);
```

Comment:
same as above

Reply:
Thank you for comments. It's not necessary so I removed it.